### PR TITLE
mds: "Failed to authpin,subtree is being exported" results in large number of blocked requests

### PR DESCRIPTION
**Pull request by orange30**
A problem: the cluster(octopus 15.2.16) has large numbers of blocked requests. The error associated with the block is:

    2023-01-02T15:59:10.078+0800 7f55b1734700  0 log_channel(cluster) log [WRN] : slow request 15364.865004 seconds old, received at 2023-01-02T11:43:05.214763+0800: client_request(client.450609:59264338 lookup                #0x40004e86a72/halo_ce_grace_F4284.spec.pt 2023-01-02T11:43:05.153256+0800 caller_uid=0, caller_gid=0{}) currently failed to authpin, subtree is being exported
     2023-01-02T15:59:10.078+0800 7f55b1734700  0 log_channel(cluster) log [WRN] : slow request 15360.774800 seconds old, received at 2023-01-02T11:43:09.304967+0800: client_request(client.450609:59265051 lookup                #0x40004e86a72/halo_ce_grace_F3233.wav 2023-01-02T11:43:09.243256+0800 caller_uid=0, caller_gid=0{}) currently failed to authpin, subtree is being exported

Eventually, many requests are blocked for hours. We can restore the cluster by restarting the affected MDS. 

The valuable log：

    2023-01-02T18:38:32.319+0800 7f55b3738700 10 mds.11.mig show_exporting  exporting to 8: (6) warning 0x40004e86a72.001001100* [dir 0x40004e86a72.001001100* /data/46f/732/03237764b1b2b824550ff4e750/data/vits_data/         generate/wavs/ [2,head] auth{0=2,1=1,2=1,3=1,6=1,8=2,10=1} v=116159 cv=116158/116158 dir_auth=11,11 state=1610875907|complete|frozentree|exporting f(v1690 m2022-12-16T15:05:36.373759+0800 1929=1929+0) n(v900 rc2022-12-16T15:05:  36.373759+0800 b822676763 1929=1929+0) hs=1929+0,ss=0+0 | ptrwaiter=1 request=0 child=1 frozen=1 subtree=1 importing=0 replicated=1 dirty=1 waiter=1 authpin=0 0x563e5ee71200]
    2023-01-02T18:38:33.103+0800 7f55b3738700 10 mds.11.mig show_exporting  exporting to 8: (6) warning 0x40004e86a72.001001100* [dir 0x40004e86a72.001001100* /data/46f/732/03237764b1b2b824550ff4e750/data/vits_data/         generate/wavs/ [2,head] auth{0=2,1=1,2=1,3=1,6=1,8=2,10=1} v=116159 cv=116158/116158 dir_auth=11,11 state=1610875907|complete|frozentree|exporting f(v1690 m2022-12-16T15:05:36.373759+0800 1929=1929+0) n(v900 rc2022-12-16T15:05:  36.373759+0800 b822676763 1929=1929+0) hs=1929+0,ss=0+0 | ptrwaiter=1 request=0 child=1 frozen=1 subtree=1 importing=0 replicated=1 dirty=1 waiter=1 authpin=0 0x563e5ee71200]

After reading the code about migrating，I think the reason is:
When one or more CEPH_SESSION_FLUSHMSG or MSG_MDS_EXPORTDIRNOTIFY msgs is lost, for example, because session is reseted or underlying connection is reconected(I think we belong to this situation); the dir will donot export anymore, this dir will be freezed forever. 

In this commit, after mds_freeze_tree_timeout(dafault:30s), we will try to cancel the export of this dir. 

EDIT:
For state EXPORT_NOTIFYING: when one or more MSG_MDS_EXPORTDIRNOTIFY msgs are lost，it will hang too, but the fix is not cancel the export of this dir.

Fixes: https://tracker.ceph.com/issues/58617
Signed-off-by: zhikuodu <duzhk@qq.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by orange30** [comment]
@vshankar @lxbsz Please help to review it, thanks！

**Comment by vshankar** [comment]
> @vshankar @lxbsz Please help to review it, thanks！

This fix really needs to start with a test case first. If one of the dir export messages are lost causing a subdir to remain frozen, it should be easy to replicate in test by adding a config that drops (some of the) dir export messages between the MDSs.

**Comment by orange30** [comment]
> > @vshankar @lxbsz Please help to review it, thanks！
> 
> This fix really needs to start with a test case first. If one of the dir export messages are lost causing a subdir to remain frozen, it should be easy to replicate in test by adding a config that drops (some of the) dir export messages between the MDSs.
[EXPORT_WARNING log1.txt](https://github.com/ceph/ceph/files/10565529/EXPORT_WARNING.log1.txt)

I agree that it should be not difficult to reproduce, and I will reproduce it later.

Add a little extra information:
1，The cluster has more than five hundred clients. A lot of clients are POD in k8s. 
2, I think our underlying network is not reliable, it has cause some mds switch to standby several times.
3, The dir hangs on EXPORT_WARNING state long time. Please see the log：《EXPORT_WARNING log1.txt》
4, From the code, I think, the only case CEPH_SESSION_FLUSHMSG or MSG_MDS_EXPORTDIRNOTIFY messages is lost can cause cluster hang in state EXPORT_WARNING.
5, If the mds sended CEPH_SESSION_FLUSHMSG message to a client which is a POD, let's name it clientA, if the clientA POD restart before it send ack, the mds will hang in EXPORT_WARNING state. ( possibly )

By the way, I am tring to give a new commit for the case cluster hang in EXPORT_WARNING state, but it need some discussion for help.

**Comment by vshankar** [comment]
We need https://github.com/ceph/ceph/pull/41969 test export killpoints.

**Comment by orange30** [comment]
For the case cluster hang in EXPORT_WARNING state, how to reproduce it:

version 15.2.16
two mds: mds0 and mds1 
one client

    Step 1: change the code and rebuild ceph-mds and use the new ceph-mds to reboot mds0.
      The main change looks like:
		void Server::flush_session(Session *session, MDSGatherBuilder& gather) {
		  if (!session->is_open() ||
			  !session->get_connection() ||
			  !session->get_connection()->has_feature(CEPH_FEATURE_EXPORT_PEER)) {
			return;
		  }
		  version_t seq = session->wait_for_flush(gather.new_sub());
		  if( g_conf()->mds_freeze_hang_test == 0 )
			dout(0) << "flush_session to: " << session->info.inst << dendl;
		  else if( g_conf()->mds_freeze_hang_test == 1 )
			dout(0) << "NOT flush_session to: " << session->info.inst << dendl;

		  if(g_conf()->mds_freeze_hang_test == 0)
			mds->send_message_client(make_message<MClientSession>(CEPH_SESSION_FLUSHMSG, seq), session);
		}
    Then, set mds0's mds_freeze_hang_test to be 1, the command looks like:
	    ceph tell mds.0 config set mds_freeze_hang_test 1

    Step 2:
    On client, execute command to export a dentry from mds0 to mds1, something like:
	    setfattr -n ceph.dir.pin -v 1 /mnt/kernelcf/test11/t1/t2/

    Then execute a command to access the dentry, something like:
        ll /mnt/kernelcf/test11/t1/t2/
	This command will hang, and the mds0's log will print that, look like:
	    2023-02-02T17:36:20.255+0800 7f195e422700  0 log_channel(cluster) log [WRN] : 1 slow requests, 0 included below; oldest blocked for > 577.952765 sec

If we use the commit https://github.com/ceph/ceph/pull/49940/commits/3209d31a16f9583c34e648fb700ed4abd7db4ee8, after about 30s, the dentry can access again.

I will reproduce the case cluster hang in EXPORT_NOTIFYING state and check the second commit later.
@vshankar

**Comment by vshankar** [comment]
@orange30 please use `mds:` prefix in commit messages.

**Comment by orange30** [comment]
> @orange30 please use `mds:` prefix in commit messages.

Done. 
I am sorry for the above mistake of git command, please help to remove the label "documentation", thanks.

**Comment by orange30** [comment]
For the case cluster hangs in EXPORT_NOTIFYING state, how to reproduce it:

version 15.2.16
three mds: mds0, mds1 and mds2 (The minimum number of mds is three.) 
one client

	Step 1: change the code, rebuild ceph-mds and use the new ceph-mds bin to reboot mds0.
	  The main change looks like:
		bool MDCache::dump_inode(Formatter *f, uint64_t number) {
		  ......
		  //in->dump(f, CInode::DUMP_DEFAULT | CInode::DUMP_PATH);   //original
		  in->dump(f, CInode::DUMP_DEFAULT | CInode::DUMP_PATH | CInode::DUMP_DIRFRAGS);   // main change 1 
		  //Using DUMP_DIRFRAGS, we can dump the inode's dir's elaborate information
		  ......
		}
		
		void Migrator::export_logged_finish(CDir *dir)
		{
		  ......
		  for (set<mds_rank_t>::iterator p = stat.notify_ack_waiting.begin();
			   p != stat.notify_ack_waiting.end();
			   ++p) {
			auto notify = make_message<MExportDirNotify>(dir->dirfrag(), stat.tid, true,
				pair<int,int>(mds->get_nodeid(), stat.peer),
				pair<int,int>(stat.peer, CDIR_AUTH_UNKNOWN));

			for (set<CDir*>::iterator i = bounds.begin(); i != bounds.end(); ++i)
			  notify->get_bounds().push_back((*i)->dirfrag());

			if( g_conf()->mds_freeze_hang_test2 == 0 )
			  dout(0) << "export_logged_finish send message to: " << *p << dendl;
			else
			  dout(0) << "export_logged_finish NOT send message to: " << *p << dendl;

			if(g_conf()->mds_freeze_hang_test2 != *p)   // main change 2; we donot send msg to the value of mds_freeze_hang_test2 for test.
			  mds->send_message_mds(notify, *p);
		  }
		  ...... 
		}

	  (1)execute the commands on client:
		setfattr -n ceph.dir.pin -v 1 /mnt/kernelcf/test11/t1
		setfattr -n ceph.dir.pin -v 2 /mnt/kernelcf/test11/t1
		setfattr -n ceph.dir.pin -v 0 /mnt/kernelcf/test11/t1
	  Then we can see the auth of the dir of /mnt/kernelcf/test11/t1 is mds0; simultaneously mds1 and mds2 has it's replicas, something like:
		ceph tell mds.0 dump inode 1099829522947  # 1099829522947 is the ino of /mnt/kernelcf/test11/t1
		......
		"dirfrags": [
			{
				"path": "/test11/t1",
			  ......
				"is_auth": true,
				"auth_state": {
					"replicas": {
						"1": 3,
						"2": 3
					}
				},
				"replica_state": {
					"authority": [
						0,
						-2
					],
					"replica_nonce": 0
				},
	  (2)set mds0's mds_freeze_hang_test2 to be 2, the command looks like:
		ceph tell mds.0 config set mds_freeze_hang_test2 2

	Step 2:
	  On client, execute command to export a dentry from mds0 to mds1, something like:
		setfattr -n ceph.dir.pin -v 1 /mnt/kernelcf/test11/t1

	  Then execute a command to access the dentry, something like:
		ll /mnt/kernelcf/test11/t1
	  This command will hang, and the mds0's log will print that, look like:
		2023-02-03T16:34:02.103+0800 7f70f7800700  0 mds.0.mig export_logged_finish export_logged_finish NOT send message to: 2
		......
		2023-02-03T16:45:41.506+0800 7f70fc009700  0 log_channel(cluster) log [WRN] : 1 slow requests, 0 included below; oldest blocked for > 694.689755 secs
		2023-02-03T16:45:46.507+0800 7f70fc009700  0 log_channel(cluster) log [WRN] : 1 slow requests, 0 included below; oldest blocked for > 699.689820 secs

If we use the commit https://github.com/ceph/ceph/pull/49940/commits/3d9e6826cf6260b3b2445f6c08eb5f0387033c86, after about 30s, the dentry can access again.
@vshankar

File: src/mds/Migrator.cc [hunk 1]
```diff
@@ -219,12 +219,25 @@ void Migrator::find_stale_export_freeze()
    * - freeze subtree A
    * - client request tries authpinning items in subtree A
    *   (wait because subtree A is freezing)
+   *
+   *
+   *- For state EXPORT_WARNING:
+   *- When one or more CEPH_SESSION_FLUSHMSG or MSG_MDS_EXPORTDIRNOTIFY msgs
```
Comment (lxbsz) [inline src/mds/Migrator.cc:225]: From the code won't the `EXPORT_WARNING` state will wait for `MSG_MDS_EXPORTDIRNOTIFY` only ? How losing the `CEPH_SESSION_FLUSHMSG` could let the exporter be stuck in `EXPORT_WARNING` state ?

**Comment by lxbsz** [comment]
When exporting started and was trying to notify the replicas, which will set the exporter to the warning state. If the replica MDSs were doing failover or something else, it could fail to handle the notify msg and then would let the exporter be stuck.

**Comment by orange30** [comment]
https://github.com/ceph/ceph/pull/49940#discussion_r1096864606

1,In export_frozen():
  (1)The export state will change to be EXPORT_PREPPING.
  (2)It send CEPH_SESSION_FLUSHMSG message to "export_client_set".
  (3)If the "export_client_set" is not empty (in other words, "gather.has_subs()" is true), MDS_RANK_NONE was insert to export_state->second->warning_ack_waiting:
      it->second.warning_ack_waiting.insert(MDS_RANK_NONE);   //from export_frozen()
2, After the exporter get the MSG_MDS_EXPORTDIRPREPACK message, it will call handle_export_prep_ack. In handle_export_prep_ack:
   (1)It sends MSG_MDS_EXPORTDIRNOTIFY message to the mds who has replica of the dir.
   (2)The state will change to be EXPORT_WARNING.
3, export_go() is the only function for exporter to switch EXPORT_WARNING to the next one( EXPORT_EXPORTING ).  
4, The only two conditions for export_go() can be called is that:
   (1)it->second.state == EXPORT_WARNING
   (2)it->second.warning_ack_waiting.empty()

Then the key point is that:
    Only all the CEPH_SESSION_FLUSHMSG_ACK has got, MDS_RANK_NONE can be erased from  export_state->second->warning_ack_waiting():
	    gather.set_finisher(new C_M_ExportSessionsFlushed(this, dir, it->second.tid));   //from export_frozen()

I donot know why my comment is always in Pending state if i reply the link directly, so I comment here. 
@lxbsz

**Comment by orange30** [comment]
> When exporting started and was trying to notify the replicas, which will set the exporter to the warning state. If the replica MDSs were doing failover or something else, it could fail to handle the notify msg and then would let the exporter be stuck.

If the exporter mds is in warning state and the replica MDSs fail, the mon will know that, and the exporter mds will call handle_mds_failure_or_stop().
In handle_mds_failure_or_stop(), it will erase the replica mds from warning_ack_waiting, and check if it can switch to next state.
https://github.com/ceph/ceph/blob/1e8aa0c17d184ee6d46c6533579156ac8baaa534/src/mds/Migrator.cc#L442
From the code I think the case replica mds fails looks ok.

**Comment by vshankar** [comment]
Thank you for the update @orange30. I"ll have a look today.

**Comment by lxbsz** [comment]
> > When exporting started and was trying to notify the replicas, which will set the exporter to the warning state. If the replica MDSs were doing failover or something else, it could fail to handle the notify msg and then would let the exporter be stuck.
> 
> If the exporter mds is in warning state and the replica MDSs fail, the mon will know that, and the exporter mds will call handle_mds_failure_or_stop(). In handle_mds_failure_or_stop(), it will erase the replica mds from warning_ack_waiting, and check if it can switch to next state.
> 
> https://github.com/ceph/ceph/blob/1e8aa0c17d184ee6d46c6533579156ac8baaa534/src/mds/Migrator.cc#L442
> 
> 
> From the code I think the case replica mds fails looks ok.

Yeah, correct. So the clients failure should be the issue here.

**Comment by orange30** [comment]
> 
> Yeah, correct. So the clients failure should be the issue here.

Most likely the clients failure has caused the stuck in warning state.
But from the code, I think another possibility of loss of MSG_MDS_EXPORTDIRNOTIFY messages cannot be completely ruled out.

Success in send_message_mds() does not guarantee Message delivery to the peer mds, only success in queueing the message in ProtocolV2::out_queue.
    https://github.com/ceph/ceph/blob/1e8aa0c17d184ee6d46c6533579156ac8baaa534/src/msg/async/ProtocolV2.cc#L457

For example:
    ProtocolV2::handle_session_reset 
	->ProtocolV2::reset_session
	->ProtocolV2::discard_out_queue
	The ProtocolV2::out_queue will be cleared, and the messages in ProtocolV2::out_queue will be lost.

**Comment by vshankar** [comment]
> > Yeah, correct. So the clients failure should be the issue here.
> 
> Most likely the clients failure has caused the stuck in warning state.

The missed FLUSHACK messaged issue seems problematic. @lxbsz should we include such message drops in killpoint tests?

> But from the code, I think another possibility of loss of MSG_MDS_EXPORTDIRNOTIFY messages cannot be completely ruled out.
> 
> Success in send_message_mds() does not guarantee Message delivery to the peer mds, only success in queueing the message in ProtocolV2::out_queue.
> 
> https://github.com/ceph/ceph/blob/1e8aa0c17d184ee6d46c6533579156ac8baaa534/src/msg/async/ProtocolV2.cc#L457
> 
> For example: ProtocolV2::handle_session_reset ->ProtocolV2::reset_session ->ProtocolV2::discard_out_queue The ProtocolV2::out_queue will be cleared, and the messages in ProtocolV2::out_queue will be lost.

Are you saying this state isn't handled when the exporter mds tries to cleanup a failed peer (importer)?

**Comment by lxbsz** [comment]
> > > Yeah, correct. So the clients failure should be the issue here.
> > 
> > 
> > Most likely the clients failure has caused the stuck in warning state.
> 
> The missed FLUSHACK messaged issue seems problematic. @lxbsz should we include such message drops in killpoint tests?

No, it doesn't.

Currently all the killpoint tests are between exporter and importer MDSs and the client is not involved.

IMO this is not a killpoint kind test, we can add this test case together in this PR.

**Comment by orange30** [comment]
> Are you saying this state isn't handled when the exporter mds tries to cleanup a failed peer (importer)?

What I understand is that:
1, Success in send_message_mds() does not guarantee Message delivery to the peer mds.
2, Obviously, the relation between mds peer has several levels:
   AsyncMessenger (let's name it level 1)
   AsyncConnection  (let's name it level 2)
   socket of linux kernel(used for epool)  (let's name it level 3)
3, out_queue belongs to ProtocolV2, and ProtocolV2 has a AsyncConnection.

But I donot sure that:
   If the level3 socket has error, do the ProtocolV2::out_queue will be cleared?
   If the level2 connection has error, do the ProtocolV2::out_queue will be cleared?
   Do the above errors can happen after the mds has sended the MSG_MDS_EXPORTDIRNOTIFY messages in warnning state?

I need to do more code reading and testing. It's a challenge for me. :)

**Comment by vshankar** [comment]
> > > > Yeah, correct. So the clients failure should be the issue here.
> > > 
> > > 
> > > Most likely the clients failure has caused the stuck in warning state.
> > 
> > 
> > The missed FLUSHACK messaged issue seems problematic. @lxbsz should we include such message drops in killpoint tests?
> 
> No, it doesn't.
> 
> Currently all the killpoint tests are between exporter and importer MDSs and the client is not involved.
> 
> IMO this is not a killpoint kind test, we can add this test case together in this PR.

Its not really a killpoint test, but can a session reset b/w the client and the mds cause this issue? I'm thinking if the dir export can possibly continue post reconnect.

**Comment by lxbsz** [comment]
> > > > > Yeah, correct. So the clients failure should be the issue here.
> > > > 
> > > > 
> > > > Most likely the clients failure has caused the stuck in warning state.
> > > 
> > > 
> > > The missed FLUSHACK messaged issue seems problematic. @lxbsz should we include such message drops in killpoint tests?
> > 
> > 
> > No, it doesn't.
> > Currently all the killpoint tests are between exporter and importer MDSs and the client is not involved.
> > IMO this is not a killpoint kind test, we can add this test case together in this PR.
> 
> Its not really a killpoint test, but can a session reset b/w the client and the mds cause this issue? I'm thinking if the dir export can possibly continue post reconnect.

In case after the `MDS` sending out a message to client,  and just before sending the ack out the client crashes or the network connection is blocked or disconnected, even the peer `MDS` daemon could receive the reset notification but it won't cancel the exporting. 

This could happen when the clients are in the containers and the netns brx interface could be disabled in anytime as the cephfs qa tests do.

In the reconnect case, not like the normal client requests the `_FLUSHMSG_ACK` ack msg won't be resent by clients, so the exporting won't continue IMO.

**Comment by vshankar** [comment]
> In the reconnect case, not like the normal client requests the `_FLUSHMSG_ACK` ack msg won't be resent by clients, so the exporting won't continue IMO.

Yeh, I know that. What I meant was can the MDS resend the SESSION_FLUSHMSG after reconnect (the mds also expect the seq number for the flush message though)? The MDS records that its waiting for a SESSION_FLUSHMSG_ACK in the client session. The client obviously resets a bunch of seq numbers it was using before the session reset, but the seq number used for FLUSHMSG, the client just sends it back in FLUSHMSG_ACK - can the MDS possibly be a bit more smart in continuing the export post reconnect?

**Comment by lxbsz** [comment]
> > In the reconnect case, not like the normal client requests the `_FLUSHMSG_ACK` ack msg won't be resent by clients, so the exporting won't continue IMO.
> 
> Yeh, I know that. What I meant was can the MDS resend the SESSION_FLUSHMSG after reconnect (the mds also expect the seq number for the flush message though)? The MDS records that its waiting for a SESSION_FLUSHMSG_ACK in the client session. The client obviously resets a bunch of seq numbers it was using before the session reset, but the seq number used for FLUSHMSG, the client just sends it back in FLUSHMSG_ACK - can the MDS possibly be a bit more smart in continuing the export post reconnect?
> 
> EDIT: the push seq is incremented for the session by the mds in send_message_client_counted, so we can't just blindly use that post reconnect.

Yeah, Sounds reasonable.

Checked the commit history about the `FLUSHMSG`, Zheng added this just want to make sure the open and export events order was guaranteed and the client did nothing except sending the ack back immediately. 

NOTE: Patrick added one patch to flush the `session->release` in the `FLUSHMSG` request in client side.

And after reconnecting we can just replace the stale waiter for old `FLUSHMSG_ACK` with a fresh new one and then continue the exporting.

**Comment by orange30** [comment]
> > In the reconnect case, not like the normal client requests the `_FLUSHMSG_ACK` ack msg won't be resent by clients, so the exporting won't continue IMO.
> 
> Yeh, I know that. What I meant was can the MDS resend the SESSION_FLUSHMSG after reconnect (the mds also expect the seq number for the flush message though)? The MDS records that its waiting for a SESSION_FLUSHMSG_ACK in the client session. The client obviously resets a bunch of seq numbers it was using before the session reset, but the seq number used for FLUSHMSG, the client just sends it back in FLUSHMSG_ACK - can the MDS possibly be a bit more smart in continuing the export post reconnect?
> 
> EDIT: the push seq is incremented for the session by the mds in send_message_client_counted, so we can't just blindly use that post reconnect.

Another problem for resending _FLUSHMSG_ACK:
1, In flush_client_sessions()->flush_session()->wait_for_flush(), it adds the gather.new_sub()+push_seq to Session::waitfor_flush.
   For example:
   (1)For dir1's export, we add new_sub()+push_seq in SessionA::waitfor_flush (SessionA belongs to clientA).
   (2)For dir2's export, we add new_sub()+push_seq in SessionA::waitfor_flush.
   ...
   (3)For dirN's export, we add new_sub()+push_seq in SessionA::waitfor_flush.
2, Then, for example, if dir2 need resend _FLUSHMSG_ACK message, from the data in Session::waitfor_flush we cannot find which new_sub()+push_seq belongs to dir2.
3, If we change the code and store dir+new_sub()+push_seq in in Session::waitfor_flush, there will be another problem:
   ----The _FLUSHMSG_ACK message donot have any information which dir it belongs to, how we delete data from Session::waitfor_flush when we get a _FLUSHMSG_ACK?

**Comment by lxbsz** [comment]
> > > In the reconnect case, not like the normal client requests the `_FLUSHMSG_ACK` ack msg won't be resent by clients, so the exporting won't continue IMO.
> > 
> > 
> > Yeh, I know that. What I meant was can the MDS resend the SESSION_FLUSHMSG after reconnect (the mds also expect the seq number for the flush message though)? The MDS records that its waiting for a SESSION_FLUSHMSG_ACK in the client session. The client obviously resets a bunch of seq numbers it was using before the session reset, but the seq number used for FLUSHMSG, the client just sends it back in FLUSHMSG_ACK - can the MDS possibly be a bit more smart in continuing the export post reconnect?
> > EDIT: the push seq is incremented for the session by the mds in send_message_client_counted, so we can't just blindly use that post reconnect.
> 
> Another problem for resending _FLUSHMSG: 1, In flush_client_sessions()->flush_session()->wait_for_flush(), it adds the gather.new_sub()+push_seq to Session::waitfor_flush. For example: (1)For dir1's export, we add new_sub()+push_seq in SessionA::waitfor_flush (SessionA belongs to clientA). (2)For dir2's export, we add new_sub()+push_seq in SessionA::waitfor_flush. ... (3)For dirN's export, we add new_sub()+push_seq in SessionA::waitfor_flush. 2, Then, for example, if dir2 need resend _FLUSHMSG message, from the data in SessionA::waitfor_flush we cannot find which new_sub()+push_seq belongs to dir2. 3, If we change the code and store dir+new_sub()+push_seq in in Session::waitfor_flush, there will be another problem: ----The _FLUSHMSG_ACK message donot have any information which dir it belongs to, how we delete data from Session::waitfor_flush when we get a _FLUSHMSG_ACK?
> 
> EDIT: when dir2 need resend _FLUSHMSG message, we also donot know which sessions we should send the _FLUSHMSG? (gather donot have enough information for resend )

The sessions will be reconnected and It should send the requests to the same sessions.


We can just wake up all the waiters in the sessions after reconnecting is done and then in the callbacks, for example `C_M_ExportSessionsFlushed`, to resend a `FLUSHMSG` request to clients with new push seq.

Actually the push seq isn't a matter.

**Comment by vshankar** [comment]
> > > > In the reconnect case, not like the normal client requests the `_FLUSHMSG_ACK` ack msg won't be resent by clients, so the exporting won't continue IMO.
> > > 
> > > 
> > > Yeh, I know that. What I meant was can the MDS resend the SESSION_FLUSHMSG after reconnect (the mds also expect the seq number for the flush message though)? The MDS records that its waiting for a SESSION_FLUSHMSG_ACK in the client session. The client obviously resets a bunch of seq numbers it was using before the session reset, but the seq number used for FLUSHMSG, the client just sends it back in FLUSHMSG_ACK - can the MDS possibly be a bit more smart in continuing the export post reconnect?
> > > EDIT: the push seq is incremented for the session by the mds in send_message_client_counted, so we can't just blindly use that post reconnect.
> > 
> > 
> > Another problem for resending _FLUSHMSG: 1, In flush_client_sessions()->flush_session()->wait_for_flush(), it adds the gather.new_sub()+push_seq to Session::waitfor_flush. For example: (1)For dir1's export, we add new_sub()+push_seq in SessionA::waitfor_flush (SessionA belongs to clientA). (2)For dir2's export, we add new_sub()+push_seq in SessionA::waitfor_flush. ... (3)For dirN's export, we add new_sub()+push_seq in SessionA::waitfor_flush. 2, Then, for example, if dir2 need resend _FLUSHMSG message, from the data in SessionA::waitfor_flush we cannot find which new_sub()+push_seq belongs to dir2. 3, If we change the code and store dir+new_sub()+push_seq in in Session::waitfor_flush, there will be another problem: ----The _FLUSHMSG_ACK message donot have any information which dir it belongs to, how we delete data from Session::waitfor_flush when we get a _FLUSHMSG_ACK?
> > EDIT: when dir2 need resend _FLUSHMSG message, we also donot know which sessions we should send the _FLUSHMSG? (gather donot have enough information for resend )
> 
> The sessions will be reconnected and It should send the requests to the same sessions.
> 
> We can just wake up all the waiters in the sessions after reconnecting is done and then in the callbacks, for example `C_M_ExportSessionsFlushed`, to resend a `FLUSHMSG` request to clients with new push seq.
> 
> Actually the push seq isn't a matter.

Using a new seq number would suffice.

**Comment by orange30** [comment]
......
> > Actually the push seq isn't a matter.
> 
> Using a new seq number would suffice.

Some question for help:
1, What does the "reconnect case" mean here? From the Server::handle_client_session level, it donot have session_reset case. I am reading the code about ProtocolV1 and ProtocolV2 which is under AsyncConnection, but now I donot know how to make a reset for session from kernel client.
2, Does the kernel client support the ProtocolV2? If it support, which linux version?
3, I think: If we add a waiter to the session when we send a FLUSHMSG message. Then when we get a FLUSHMSG_ACK message, we should delete a waiter simultaneously, or if the session is reconneted, it will make a mistake.

**Pull request closed without merging**